File: MapReduce and Spark/scala/task3.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s._
import org.json4s.jackson.JsonMethods._

import java.io.{BufferedWriter, File, FileWriter}
import scala.util.parsing.json.JSONObject
// My prediction: My Scala code will run faster then my Spark code on a large dataset on Vocareum:
//
// Reason : The grading dataset is about 5GB and it is smaller than recent main memory size.
// When I run code on Scala, it will store all the items into the main memory and sort them
// However, when I run code with Spark, as the items are divided into multiple partitions, it has to do reshuffling process to sort the data.
// This will make Scala Code faster than Spark code.

object task3 {
  def main(args: Array[String]): Unit = {

    //     get argument
    //    val review_filepath = "/Users/jaeyoungkim/IdeaProjects/hw1/src/main/scala/test_review.json"

    val review_filepath = args(0)
    val business_filepath = args(1)
    val output_filepath_question_a = args(2)
    val output_filepath_question_b = args(3)


    def reduce_ftn(first : (Int, Double), second: (Int, Double)):(Int, Double)={
      return (first._1 + second._1, first._2 + second._2)
    }
    def map_ftn(x : (String, (Int, Double))): (Double, String)={
      return (x._2._2/x._2._1, x._1)  //avg, count
    }

    def ftn_(line: (Double, List[String])): List[(String, Double)] = {
      val num = line._1
      val list_of_names = line._2
      var res: List[(String, Double)] = List()
      for (w <- 0 to list_of_names.size - 1) {
        //        println((num, list_of_names(w)))
        //        res :+ (num, list_of_names(w))
        res = res :+ ((list_of_names(w), num))
      }
      return res
    }
    def ftn2(line: String, col_name : String): String = {
      val data = parse(line)
      val date = data \ col_name
      val JString(res) = date
      return res
    }
    def ftn4(line: String, col_name : String): Double = {
      val data = parse(line)
      val date = data \ col_name
      val JDouble(res) = date
      return res.toDouble
    }

    val t2 = System.nanoTime
    val spark = new SparkConf().setAppName("task3").setMaster("local[*]")
    val sc = new SparkContext(spark)

    val sample1RDD = sc.textFile(review_filepath)
    val sample2RDD = sc.textFile(business_filepath)

    //m1
    val t1 = System.nanoTime
    val sample1RDD1 = sc.textFile(review_filepath)
    val sample2RDD1 = sc.textFile(business_filepath)


    val tmp1 = sample1RDD1.map(line => (ftn2(line, "business_id"), ftn4(line, "stars")))
    val tmp21 = sample2RDD1.map(line => (ftn2(line, "business_id"), ftn2(line, "city")))

    val joined1 = tmp1.join(tmp21)

    val joined21 = joined1.map(line => (line._2._2,(1,line._2._1)))

    println(joined21.collect()(1))

    val joined31 = joined21.reduceByKey(reduce_ftn)





    val question_b_tmp = joined31.map(x => map_ftn(x)).collect()
    println(question_b_tmp(1))

    val question_b_tmp_1 = question_b_tmp.sortBy(question_b_tmp => (-question_b_tmp._1, question_b_tmp._2)).slice(0,10)
    val duration1 = (System.nanoTime - t1) / 1e9d

    ////////////////

    val tmp = sample1RDD.map(line => (ftn2(line, "business_id"), ftn4(line, "stars")))
    val tmp2 = sample2RDD.map(line => (ftn2(line, "business_id"), ftn2(line, "city")))
    println(tmp2.collect())
    println("################################################")
    val joined = tmp.join(tmp2)

    println("################################################")
    val joined2 = joined.map(line => (line._2._2,(1,line._2._1)))

    println(joined2.collect()(1))



    val joined3 = joined2.reduceByKey(reduce_ftn)



    val question_a_tmp = joined3.map(x => map_ftn(x)).groupByKey().sortByKey(false).mapValues(v => v.toList.sortWith(_ < _)).flatMap(x => ftn_(x)).take(10)
    println(question_a_tmp(1))
    val duration2 = (System.nanoTime - t2) / 1e9d
    println(duration2)

    val file = new File(output_filepath_question_a)
    val bw = new BufferedWriter(new FileWriter(file))
    bw.write("city, stars"+"\n")
    for (w <- 0 to question_a_tmp.size-1){

      bw.write(question_a_tmp(w)._1 +','+ question_a_tmp(w)._2.toString+ "\n")

    }
    bw.close()

    val answer = Map("m1" -> duration1, "m2" -> duration2)

    // FileWriter
    val file1 = new File(output_filepath_question_b)
    val bw1 = new BufferedWriter(new FileWriter(file1))
    bw1.write(JSONObject(answer).toString())
    bw1.close()


  }
}

File: 1. MapReduce and Spark/python/task3.py

# My prediction: my Python code will run faster then my Spark code on a large dataset on Vocareum:
#
# Reason : The grading dataset is about 5GB and it is smaller than recent main memory size.
# When I run code on python, it will store all the items into the main memory and sort them
# However, when I run code with Spark, as the items are divided into multiple partitions, it has to do reshuffling process to sort the data.
# This will make Python Code faster than Spark code.

import time
import json
from pyspark import SparkContext
import sys

def main(review_filepath, business_filepath, output_filepath_question_a, output_filepath_question_b):


    sc = SparkContext('local[*]', 'task3')
    start_time = time.time()

    sample1RDD = sc.textFile(review_filepath)
    sample2RDD = sc.textFile(business_filepath)

    tmp = sample1RDD.map(lambda line : (json.loads(line)["business_id"],(json.loads(line)["stars"])))
    tmp2 = sample2RDD.map(lambda line : (json.loads(line)["business_id"],(json.loads(line)["city"])))

    joined = tmp.join(tmp2)

    joined = joined.map(lambda line :(line[1][1],(1,line[1][0])))

    def ftn1(first,second):
        return (first[0]+second[0], first[1]+second[1]) ##count, sum
    joined = joined.reduceByKey(ftn1)

    def ftn2(x): ##(avg_star, city)
        return (x[1][1]/x[1][0], x[0])

    question_a_tmp = joined.map(lambda x : ftn2(x)).groupByKey().sortByKey(False).mapValues(lambda v: sorted(v)).flatMap(lambda x : [(x[0], value) for value in x[1]]).collect()
    time_m2 = (time.time() - start_time)

    ### store result of a
    with open(output_filepath_question_a, 'w') as writefile:
        writefile.writelines("city,stars\n")
        for i in question_a_tmp:
            writefile.writelines(str(i[1]) + ',' + str(i[0]) + '\n')



    ######m2
    start_time = time.time()

    sample1RDD = sc.textFile(review_filepath)
    sample2RDD = sc.textFile(business_filepath)

    tmp = sample1RDD.map(lambda line: (json.loads(line)["business_id"], (json.loads(line)["stars"])))
    tmp2 = sample2RDD.map(lambda line: (json.loads(line)["business_id"], (json.loads(line)["city"])))

    joined = tmp.join(tmp2)
    joined = joined.map(lambda line: (line[1][1], (1, line[1][0])))

    def ftn1(first, second):
        return (first[0] + second[0], first[1] + second[1])

    joined = joined.reduceByKey(ftn1)

    def ftn2(x):
        return (x[1][1] / x[1][0], x[0])

    def ftn2(x):
        return (x[1][1] / x[1][0], x[0])

    tmp = joined.map(lambda x: ftn2(x)).collect()

    tmp = sorted(tmp, key=lambda x: (-x[0], x[1]))

    if len(tmp) >= 10:
        tmp = tmp[0:10]

    time_m1 = (time.time() - start_time)

    question_b = {}
    question_b["m1"] = time_m1
    question_b["m2"] = time_m2

    with open(output_filepath_question_b, 'w') as json_file:
        json.dump(question_b, json_file)



review_filepath = sys.argv[1]
business_filepath = sys.argv[2]
output_filepath_question_a = sys.argv[3]
output_filepath_question_b = sys.argv[4]

# main(review_filepath, output_file_path, n_partition)
main(review_filepath, business_filepath, output_filepath_question_a, output_filepath_question_b)
